libc/string/string_tokpbrkstr.c

```diff
@@ -0,0 +1,1019 @@
+
+TEST(string_str, basic)
+{
+	const char loremIpsum[BUFFSIZE] = LOREM_IPSUM;
+	char *asciiStr = testdata_createCharStr(ASCII_LENGTH);
+
+	/* Standard use of strstr on arrays */
+	TEST_ASSERT_EQUAL_PTR(loremIpsum, strstr(loremIpsum, "Lorem"));
+	TEST_ASSERT_EQUAL_PTR(&loremIpsum[6], strstr(loremIpsum, "Ipsum"));
+	TEST_ASSERT_EQUAL_PTR(&loremIpsum[12], strstr(loremIpsum, "Dolor"));
+
+	/* Simple usage of strstr with ascii set */
+	TEST_ASSERT_EQUAL_PTR(asciiStr, strstr(asciiStr, asciiStr));
+	TEST_ASSERT_EQUAL_PTR(&asciiStr[ASCII_LENGTH / 4], strstr(asciiStr, &asciiStr[ASCII_LENGTH / 4]));
+	TEST_ASSERT_EQUAL_PTR(&asciiStr[ASCII_LENGTH / 3], strstr(asciiStr, &asciiStr[ASCII_LENGTH / 3]));
+	TEST_ASSERT_EQUAL_PTR(&asciiStr[ASCII_LENGTH / 2], strstr(asciiStr, &asciiStr[ASCII_LENGTH / 2]));
+
+
+	free(asciiStr);
+}
+
+
+TEST(string_str, empty_args)
+{
+	const char str[] = "Lorem";
+
+
+	char *asciiStr = testdata_createCharStr(ASCII_LENGTH);
+
+	/* Different scenarios of using empty input or output*/
+	TEST_ASSERT_EQUAL_PTR(NULL, strstr("", str));
+	TEST_ASSERT_EQUAL_STRING("", strstr("", ""));
+	TEST_ASSERT_EQUAL_PTR(str, strstr(str, ""));
+
+	/* Passing an empty string as the second parameter to get first */
+	TEST_ASSERT_EQUAL_PTR(asciiStr, strstr(asciiStr, ""));
+
+
+	free(asciiStr);
+}
+
+
+TEST(string_str, out_of_ascii)
+{
+	int i;
+
+	char *extAsciiStr = create_extAscii_set();
+
+	for (i = 1; i < sizeof(extAsciiStr); i++) {
+		TEST_ASSERT_EQUAL_PTR(&extAsciiStr[i], strstr(extAsciiStr, &extAsciiStr[i]));
+	}
+
+	free(extAsciiStr);
+}
+
+
+TEST(string_str, strstr_order)
+{
+	int i;
+	char reversStr[ASCII_LENGTH],
+		*testStr;
+
+	testStr = testdata_createCharStr(ASCII_LENGTH);
+
+	TEST_ASSERT_NOT_NULL(reversStr);
+
+	for (i = 0; i < ASCII_LENGTH - 1; i++) {
+		reversStr[i] = testStr[ASCII_LENGTH - i - 2];
+	}
+	reversStr[ASCII_LENGTH - 1] = 0;
+
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "cba"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bac"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bca"));
+
+	/* Asserting reversed string */
+	TEST_ASSERT_EQUAL_PTR(NULL, strstr(testStr, reversStr));
+
+	free(testStr);
+}
+
+
+TEST(string_str, part_of_str)
+{
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "aa"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bb"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "cc"));
+	TEST_ASSERT_EQUAL_STRING("abc", strstr("abc", "ab"));
+	TEST_ASSERT_EQUAL_STRING("bc", strstr("abc", "bc"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "ac"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "xbc"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "xab"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "aabbcc"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "aaaaaa"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bbbbbb"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "cccccc"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "ccaabb"));
+	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bbaacc"));
```

Comment:
- TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bb")); already asserted by a previous "aa" check, so we can remove it and "cc"
- TEST_ASSERT_EQUAL_STRING("abc", strstr("abc", "ab")); checks nothing more than cases in `basic` - to remove, same with "bc"

TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "aabbcc"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "aaaaaa"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bbbbbb"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "cccccc"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "ccaabb"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bbaacc"));
I think those cases are not necessary too. If it does not catch 2 same bytes it rather won't catch more.

Also I'd add `TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bcx"));`, because there is some potential for an issue.

```suggestion
TEST(string_str, partially_included)
{
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "aa"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "ac"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "xbc"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "bcx"));
	TEST_ASSERT_EQUAL_STRING(NULL, strstr("abc", "xab"));
```